scripts/count-bugs.py

```diff
@@ -56,18 +71,41 @@ def print_summary(issues):
             text=warning, count=count, field_size=size))
     print()
 
+def parse_args(argv):
+    # Guard: show the usage block for incorrect usage
+    if len(argv) not in [2, 3]:
+        print(__doc__)
+        sys.exit(1)
+    # Guard: check if the provided report file exists
+    report = argv[-1]
+    if not os.path.isfile(report):
+        print("{}: no such file".format(report))
+        sys.exit(1)
+	# Good scenario: set our maximum and return it with the log
+    max_issues = MAX_ISSUES if len(argv) == 2 else int(argv[1])
+    return max_issues, report
+
 
-def main():
-    bug_types = read_soup(sys.argv[1])
+def main(argv):
+    rcode = 0
+    max_issues, report = parse_args(argv)
+    bug_types = read_soup(report)
     total = bug_types.pop('All Bugs')
     if bug_types:
         print("Bugs grouped by type:\n")
         print_summary(bug_types)
-    print('Total: {} bugs (out of {} allowed)\n'.format(total, MAX_ISSUES))
-    if total > MAX_ISSUES:
-        print('Error: upper limit of allowed bugs is', MAX_ISSUES)
-        sys.exit(1)
 
+    print('Total: {} bugs'.format(total), end='')
+    if max_issues >= 0:
+        print(' (out of {} allowed)\n'.format(max_issues))
+        if total > max_issues:
+            print('Error: upper limit of allowed bugs is', max_issues)
+            rcode = 1
+    else:
+        print('\n')
+
+    return rcode
 
 if __name__ == '__main__':
-    main()
+    import sys
+    sys.exit(main(sys.argv))
```

Comment:
Will update.  But it's a common pattern in cases where sys is only used as the interface to the shell (args in, return-code out). 

 In these cases sys is only activated when executed as a shell script. If the file is instead used as a module, then this shell interface stays dormant and is not part of the namespace.

Reply:
Does it make a difference? `sys` module is built-in Python interpreter (so importing it does not touch the filesystem nor even uses additional memory, just exposes the symbols). If we were talking about other module, which would be dead import when file is imported and not directly executed, then I would agree.

Reply:
Was the basis of driving PEP 299, to standardize the shell interface that would hook into a `main(argv)` function automatically:

``` python
if globals().has_key("__main__"):
    import sys as __sys
    __sys.exit(__main__(__sys.argv))
```

It was rejected not because of anything wrong, but because it's _"not worth the change (in docs, user habits, etc.) and there's nothing particularly broken_."

Reply:
> built-in Python interpreter (so importing it does not touch the filesystem nor even uses additional memory, just exposes the symbols)

Yeah, it's just a minor namespace cleanup and avoids needing the code inside main to use sys.argv; instead it receive the bare argv list just like in C++; the `sys.exit(main(sys.argv))` pattern is pretty common and I tend to prefer it.

Reply:
Reverted in 69e5ea39